lib/babe/verify.go

```diff
@@ -327,62 +327,58 @@ func (b *verifier) verifyAuthorshipRight(header *types.Header) error {
 		return ErrBadSignature
 	}
 
-	// check if the producer has equivocated, ie. have they produced a conflicting block?
-	// hashes is hashes of all blocks with same block number as header.Number
-	hashes := b.blockState.GetAllBlocksAtDepth(header.ParentHash)
+	equivocated, err := b.verifyBlockEquivocation(header)
+	if err != nil {
+		return fmt.Errorf("could not verify block equivocation for header %s: %w", header.Hash(), err)
+	}
+	if equivocated {
+		return ErrProducerEquivocated
+	}
 
-	for _, currentHash := range hashes {
-		currentHeader, err := b.blockState.GetHeader(currentHash)
-		if err != nil {
-			return fmt.Errorf("failed get header %s", err)
-		}
+	return nil
+}
 
-		currentBlockProducerIndex, err := getAuthorityIndex(currentHeader)
-		if err != nil {
-			return fmt.Errorf("failed to get authority index %s", err)
-		}
+// verifyBlockEquivocation checks if given block's author has occupied corresponding slot more than once.
+// It returns true if block was equivocated.
+func (b *verifier) verifyBlockEquivocation(header *types.Header) (bool, error) {
+	author, err := getAuthorityIndex(header)
+	if err != nil {
+		return false, fmt.Errorf("failed to get authority index for %s: %w", header.Hash(), err)
+	}
 
-		if len(currentHeader.Digest.Types) == 0 {
-			return fmt.Errorf("current header missing digest")
+	currentHash := header.Hash()
+	slot, err := types.GetSlotFromHeader(header)
+	if err != nil {
+		return false, fmt.Errorf("failed to get slot for block %s: %w", currentHash, err)
+	}
+
+	blocksInSlot, err := b.blockState.GetBlocksBySlot(slot)
```

Comment:
I have a question about this method: Since `GetBlocksBySlot` retrieves finalized blocks (given that `GetBlocksBySlot` get all descendants of `GetHighestFinalisedHash`) produced on slot `X` why do we need to verify the block equivocation?

I mean, if `GetBlocksBySlot` already returns finalized blocks on slot X any other block being created on the same epoch should be invalid, right?

I believe you should change the method `GetBlocksBySlot` to rely on `BestBlock` (best block not finalized yet) to get blocks produced on a given slot and then verify equivocation

Reply:
descendants of `GetHighestFinalisedHash` blocks will not be finalised.

Reply:
lets' say `GetHighestFinalisedHash` gives you `x`. If `x` has a descendant block `d` that is finalized, that would make `d` the highest finalised block.